### ISSUE TITLE
Setup Stats Service

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#923]
Goals:

Understand the limits of our software in the wild

Plan of Attack:

A) Instrument a lot more of our code and have it send stats to segment.io

- pachd
  - all API calls and timing (e.g. `StartCommit` etc)
  - PPS -- bytes read / deleted ... lots of metrics around Pipelines and their run patterns (wrt Jobs / Pods / etc)
  - Cache hit/miss ratios
  - # of open client connections
  - # of unique users being serviced by a cluster
- FUSE
  - all FUSE calls (e.g. `open`, `create`, etc)
- Job Shim
  - File I/O
  - Network I/O
  - Socket states (open/closed/wait/etc)
  - Bytes read (per pod!) this will help validate read caching and pre-fetch usage

B) Look at the data in mixpanel and evaluate if there is a better front end to view the statistics

- Compare some other front ends
- Figure out how we'd namespace stats per cluster
- Figure out cost structure

**Comment by msteffen** [comment]
LGTM. Just to recap our verbal discussion, I think we tentatively concluded that segment/mixpanel would be a good place to keep sending this data. We also talked about metrics (cheap, lower-availability) vs monitoring (expensive, super duper high availability) and how we may eventually want to use different metrics and/or a different backend for monitoring.

**Comment by sjezewski** [comment]
Cool, thanks for the extra notes

Also, metrics can have overlapping but different goals. Those are A) a debugging tool and B) a tool to understand how our software is used and the limits at which it breaks

Using segment definitely should satisfy B and could in theory also satisfy A. In contrast, if we bundled a promtheus/grafana configuration as part of our k8s manifest, it would be a good debug tool (A) ... when we could get access to it ... but would not provide any help at all for B.

If we had to choose, I think we value B (understanding usage) over A (pure debugging tool), but in theory segment gives us both.

If we learn along the way that we won't be able to send everything to segment, we could always double report / report to different services (one for metrics and usage, one for stats for debugging).

**Comment by msteffen** [comment]
I guess I do think we might want to be a little careful about relying on mixpanel for debugging, for two reasons:

1) Primarily, if we lean on it too much, we may be incentivized to keep putting more and more granular data in there to understand a bad cluster's state, which could cause us to hit a scalability or cost wall at some point, and might also be bad for our users' privacy. I think the scalability constraint in particular may mean that inevitably, some day, we won't be able to debug a broken cluster without a customer giving us access to their monitoring in some way, and therefore I'm not sure that the benefit of circumventing that will be large or long-lived.

2) Presuming Mixpanel's reliability is actually worse than Prometheus, we may end up in a state where our ability to debug a broken cluster depends on whether Mixpanel was available and actually collected all of the information we need when the cluster broke. If, instead, we come up with some more robust debugging infrastructure from the start, we may end up in a better place and more consistently able to debug stuff

Obviously we should start with what's easy, but these were just some thoughts.

**Comment by jdoliner** [comment]
@msteffen I think these are very reasonable concerns. In general, I think we should avoid using mixpanel as a debugging tool and try to use more cluster local things for that. Having our processes build up with Mixpanel specific stuff would definitely be risky. Right now everything that gets to mixpanel also gets logged by pachd so you can recover that information by checking logs and you'll have more context with it since you can see what else is being logged around the same time. That will hopefully help to ameliorate this a bit.
I like the direction of keeping prometheus + logs focussed on things that allow people to understand and debug the cluster's state right now and having mixpanel focussed more on allowing us to understand broadly how it's being used in the wild by people.

**Comment by sjezewski** [comment]
I'm picking this up again. To re-state the use case:

"As a user, I want to see visual statistics that inform me how my jobs are running and where they're hitting walls"

Some specific stats we want to measure + graph:

- graph of datum download / run / upload times ("oh ... look ... that one datum out of my ten total had a spike in run time ... let me grab it's ID so I can debug further")
- graph of job run times ("oh yesterdays job was fine ... but today's job took 10x longer")

## Iteration 0

- standup prometheus + grafana deployment to scrape/store/display these app stats
- report a single type of app stat so that we can see it displayed in grafana

## Iteration 1

- instrument code to fill out other places where we want app stats reported
- examples include
    - detailed stats per datum and per job
    - stats per API call on the PFS side cars (maybe # of api calls / load times per call ... so we can see if a particular side car is getting too much load ... which could happen depending on the datum shape)

## Future Iterations

- Add mechanism for users to report custom app stats (See #228 )
    - We'd probably want to report/scrape app stats in flight
    - Either we expect them to write prometheus gauges in their user code ... OR perhaps
    - We have something like `/pfs/stats` which is a pipe so that we can listen for stats, and send them along via the prometheus client
- Make pachyderm a first class grafana plugin
    - Which would allow us to bypass prometheus, but more interestingly to users ...
    - Would allow output from a pipeline to be graphed by grafana ... which would be very handy for users trying to create reports / dashboards

**Comment by gabrielgrant** [comment]
I certainly agree what you're asking for would be useful, but, unless i'm missing something, it doesn't sound to me like either of the two use-case examples (datum and job runtimes) are really the type of time-series metric data that any of these tools (grafana, prometheus, graphite...or collecd/RRDtool et al) are made for. My understanding is that, fundamentally, these plot a metric's value over time (you implement a collector that gets called on a regular interval and reports a value, the tool drops that datapoint on a graph, and potentially does rollups to compress historical data by summing, or averaging, or whatever aggregation)

Traditionally, these are used for things like monitoring memory or CPU usage. Things that are "poll-able". If we wanted to, say, show the number of jobs running at a given moment over time, then this would be the tool, but i'm not sure how (and whether) to make use of them to collect these more event-based data forms you're looking for

FWIW, (as @sjezewski pointed out in #228), Button have open sourced a [Prometheus exporter for pachyderm](https://github.com/button/pachyderm_exporter). The list of metrics captured by that button collector are again things that can be polled, such as: total number of jobs in various states; total number of datums processed; total bytes uploaded and downloaded by all jobs, etc.

While interesting, these seem less useful overall (and certainly less useful for debugging problems with specific jobs or datums) than the types of specific event-based info that (seems to be) motivating this issue

It seems something like a heatmap or set of stacked bars that let you drill down to the individual item level would be better address what you're trying to achieve

**Comment by jdoliner** [comment]
Job runtimes is definitely something that makes sense to track in Prometheus. They actually have a [page on instrumentation](https://prometheus.io/docs/practices/instrumentation) that lists common gauges to take and the section for batch jobs lists runtime as one of the gauges you should definitely be taking. I think you're misunderstanding a little how Prometheus relates to time-series, it's not that the data you throw at it needs to be a time series, it can be much less structured than that, it's more that what it exposes to you, when you're querying the data, is time-series based. So you can just throw runtimes at it everytime a job completes, and from that it will be able to give you the average job runtime as a function of time, which is a time-series. That's my understanding at least.

**Comment by gabrielgrant** [comment]
Yes, if I wasn't clear enough about that: it's _very_ likely I am misunderstanding parts of what these tools do/how they work :P Regardless, I certainly do think having some of this time-series info is useful, just want to make sure that we're clear about what we are and aren't getting from them. Specifically: this gives us aggregate trends over time, but not a direct way to identify and drill into outliers (especially at the datum level). Spent a little time today sketching out some ideas of how we might want to present some of that more detailed/connected info in the dash. Maybe chat quickly about whether some options and what we think would be useful tomorrow?

**Comment by sjezewski** [comment]
OK a bunch of stats stuff has landed. Now we:

- [record application stats for pipelines for all worker pods (if enterprise is enabled)](https://github.com/pachyderm/pachyderm/pull/2856)
- [record all API calls for the pachd instance](https://github.com/pachyderm/pachyderm/pull/2875)
- [report the groupcache stats](https://github.com/pachyderm/pachyderm/pull/2880)

Remaining iterations / todos:

- support pachd sidecar API calls
  - id have to have more than one prometheus HTTP server on this pod, which requires reconfiguring how prometheus looks to discover services to scrape
- provide deployment via helm chart
  - right now we just have a make task to do deployment
  - after much discussion, we dont want to support this via `pachctl deploy ...`
  - the right way to support this is to add / configure it as part of our helm chart
- [allow hook for custom user metric reporting](https://github.com/pachyderm/pachyderm/issues/228)
